src/Admin/AdminHelper.php

```diff
@@ -52,25 +52,51 @@ class AdminHelper
     private $propertyAccessor;
 
     /**
-     * NEXT_MAJOR: Make $propertyAccessor mandatory and remove $pool.
+     * NEXT_MAJOR: Change signature for (PropertyAccessorInterface $propertyAccessor).
      */
-    public function __construct(Pool $pool, ?PropertyAccessorInterface $propertyAccessor = null)
+    public function __construct($poolOrPropertyAccessor, ?Pool $pool = null)
     {
-        $this->pool = $pool;
+        // NEXT_MAJOR: Remove this block.
+        if (!$poolOrPropertyAccessor instanceof Pool && !$poolOrPropertyAccessor instanceof PropertyAccessorInterface) {
+            throw new \TypeError(sprintf(
+                'Argument 1 passed to "%s()" must be either an instance of %s or %s, %s given.',
+                __METHOD__,
+                Pool::class,
+                PropertyAccessorInterface::class,
+                \is_object($poolOrPropertyAccessor) ? \get_class($poolOrPropertyAccessor) : \gettype($poolOrPropertyAccessor)
+            ));
+        }
 
         // NEXT_MAJOR: Remove this block.
-        if (null === $propertyAccessor) {
+        if ($poolOrPropertyAccessor instanceof Pool) {
             @trigger_error(sprintf(
-                'Omitting the argument 2 for "%s()" or passing "null" is deprecated since sonata-project/admin-bundle'
-                .' 3.x and will throw a \TypeError in version 4.0. You must pass an instance of %s instead.',
+                'Passing an instance of "%s" as the argument 1 for "%s()" is deprecated since'
+                .' sonata-project/admin-bundle 3.x and will throw a \TypeError in version 4.0.'
+                .' You MUST pass an instance of %s instead.',
+                Pool::class,
                 __METHOD__,
                 PropertyAccessorInterface::class
             ), E_USER_DEPRECATED);
 
-            $propertyAccessor = $pool->getPropertyAccessor();
+            $this->pool = $poolOrPropertyAccessor;
+            $this->propertyAccessor = $poolOrPropertyAccessor->getPropertyAccessor();
+
+            return;
         }
 
-        $this->propertyAccessor = $propertyAccessor;
+        // NEXT_MAJOR: Remove this block.
+        if (null === $pool) {
+            throw new \TypeError(sprintf(
+                'Argument 2 passed to "%s()" must be an instance of "%s", "null" given.',
+                __METHOD__,
+                Pool::class
+            ));
+        }
```

Comment:
Imho, it should be

```
if ($pool instanceof Pool) {
    trigger deprecation
}

$this->pool = pool
```

Then, the pool property should be typehint
```
/**
 * @var Pool|null
 */
protected $pool;
```

And the `getAdmin` method should be this way: 
```
public function getAdmin($code)
    {
        if (null === $this->pool) {
             throw \Exception()
        } 

         trigger deprecation

                 return $this->pool->getInstance($code);
    }

```

This allow people us and developers to use `new AdminHelper($propertyAccessor)` if we don't use the deprecated method `getAdmin`. The 4.0 upgrade will be easier.

Reply:
the problem with that I think it's how to avoid the deprecation because we have to inject both to not break BC.

Reply:
Passing a third argument "Sonata_deprecation_mute" ?

Reply:
mmm so maybe just injecting Pool as second argument and do not declare it in the constructor.

Reply:
Indeed